_posts/2021-12-09-cdk-cr.md

```diff
@@ -0,0 +1,174 @@
+        actions=[
+            "iam:PassRole"
+        ],
+        resources=[role_arn]
+    )]
+)
+```
+
+For each of the resource events — `on_create`, `on_update` and `on_delete` you can provide a `AwsSdkCall` object which defines which AWS SDK gets called.
+
+For example, to create an IoT Thing Group on the create event:
+
+```py
+on_create=cr.AwsSdkCall(
+    action="createThingGroup",
+    service="Iot",
+    parameters={
+        'thingGroupName': thing_group_name
+    },
+    physical_resource_id=cr.PhysicalResourceId.of(thing_group_name)
+)
+```
+
+This will call the [`createThingGroup` AWS SDK method](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Iot.html#createThingGroup-property) at resource creation time and pass parameters.
+
+`physical_resource_id` is a way of uniquely identifying the resource created at runtime. In most cases you would use the ARN of the object, which you can get from the response, but you may use something else.
+
+For example, you can use the [`fromResponse` helper](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.custom_resources.PhysicalResourceId.html#static-fromwbrresponseresponsepath) to specify the response path to use as the physical resource ID:
+
+```py
+cr.PhysicalResourceId.fromResponse('thingGroupArn')
+```
+
+Or you can specify your own ID based on any arbitrary string:
+
+```py
+cr.PhysicalResourceId.of('some_unique_id')
+```
+
+Here's a full example of a custom resource for AWS IoT Thing Groups:
+
+```py
+cr.AwsCustomResource(self, 'ThingGroup',
+                     policy=cr.AwsCustomResourcePolicy.from_sdk_calls(
+                         resources=cr.AwsCustomResourcePolicy.ANY_RESOURCE
+                     ),
+                     on_create=cr.AwsSdkCall(
+                         action="createThingGroup",
+                         service="Iot",
+                         parameters={
+                             'thingGroupName': thing_group_name
+                         },
+                         physical_resource_id=cr.PhysicalResourceId.fromResponse('thingGroupArn')
+                     ),
+                     on_delete=cr.AwsSdkCall(
+                         action="deleteThingGroup",
+                         service="Iot",
+                         parameters={
+                             'thingGroupName': thing_group_name
+                         },
+                     ))
+```
+
+This resource will create a thing group on resource creation and then delete the thing group on resource deletion. The `AwsCustomResource` construct handles the details of how that's implemented in lambda.
+
+## Summary
+
+The most ideal would solution would be if AWS found a way to improve the CloudFormation support across all its services, then we wouldn't need to polyfill it for services whose config we want to capture as code. Since we don't have that, the CDK Custom Resources module is a nice fallback for those that are using CDK for their infrastructure.
```

Comment:
`Ideally, AWS would found a way to improve the CloudFormation support across all its services`